src/ray/common/id.cc

```diff
@@ -180,18 +188,36 @@ TaskID TaskID::ForActorTask(const JobID &job_id, const TaskID &parent_task_id,
   std::string data = GenerateUniqueBytes(job_id, parent_task_id, parent_task_counter, 0,
                                          TaskID::kUniqueBytesLength);
   std::copy_n(actor_id.Data(), ActorID::kLength, std::back_inserter(data));
+  static uint8_t attempt_number = 0;
+  std::copy_n(&attempt_number, sizeof(attempt_number), std::back_inserter(data));
   RAY_CHECK(data.size() == TaskID::kLength);
   return TaskID::FromBinary(data);
 }
 
 TaskID TaskID::ForNormalTask(const JobID &job_id, const TaskID &parent_task_id,
                              size_t parent_task_counter) {
-  std::string data = GenerateUniqueBytes(job_id, parent_task_id, parent_task_counter, 0,
-                                         TaskID::kUniqueBytesLength);
+  TaskID task_id = Nil();
+  std::string unique_bytes = GenerateUniqueBytes(
+      job_id, parent_task_id, parent_task_counter, 0, TaskID::kUniqueBytesLength);
+  std::memcpy(task_id.id_, unique_bytes.data(), TaskID::kUniqueBytesLength);
   const auto dummy_actor_id = ActorID::NilFromJob(job_id);
-  std::copy_n(dummy_actor_id.Data(), ActorID::kLength, std::back_inserter(data));
-  RAY_CHECK(data.size() == TaskID::kLength);
-  return TaskID::FromBinary(data);
+  std::memcpy(task_id.id_ + TaskID::kUniqueBytesLength, dummy_actor_id.Data(),
+              ActorID::kLength);
+  // Zero out the task attempt number.
+  task_id.id_[TaskID::kUniqueBytesLength + ActorID::kLength] = 0;
+  return task_id;
+}
+
+TaskID TaskID::ForExecutionAttempt(const TaskID &task_id, uint64_t attempt_number) {
+  TaskID task_id_copy = task_id;
+  if (attempt_number >= std::numeric_limits<uint8_t>::max()) {
```

Comment:
Why not use int64_t* and allow it to overflow into the unique bits as before?

Reply:
It turns out to be tricky because of endianness (see above conversation with @jjyao). If you have an idea of how to do it in a portable/simple way, let me know.

Reply:
Can't we just assume little endian? There's no real architecture that's big endian.

Reply:
Hmm I think this is a little sketchy... it appears that some architectures default to little endian but do allow you to switch.

I could also just change this back to overflowing into the unique bits and statically detect the endianness. What do you think?

Reply:
If we want to assume little endian, one thing we could do is crash the program if it's not using https://github.com/abseil/abseil-cpp/blob/master/absl/base/internal/endian.h

Reply:
Yea, I think we need to make the assumption explicit.

Reply:
Thanks for the tip, I'll try that!